[숫자 야구 게임] FE/복수 미션 제출합니다.

Lennon57 opened the pull request:
깃이 미숙해 테스트 commit이 많으며 11/28일부터 실제 커밋입니다.
코드리뷰를 열어주셔셔 감사합니다. 많이 배우겠습니다.
피드백듣고 다음 미션부터 적용하고 코드를 발전시키겠습니다.

Lennon57 committed:
랜덤숫자 생성 및 서로 다른 숫자 비교 후 console 출력

Lennon57 committed:
커밋 테스트

Lennon57 committed:
테스트

Lennon57 committed:
readme수정

Lennon57 committed:
README 구현할 기능 목록  작성

Lennon57 committed:
입력 기능 구현
- 입력 예외처리(3글자,숫자,중복,0제외)
- 예외됐을 때 alert 및 다시입력

Lennon57 committed:
MissionUtils 라이브러리 사용해 랜덤 숫자 3개 생성

Lennon57 committed:
play메서드에서 사용자가 입력한 수와
컴퓨터의 수를 비교한다.
비교한 값을 스트라이크, 볼, 낫싱 문자열로 리턴한다.

Lennon57 committed:
README 기능 목록 수정 및 보완

Lennon57 committed:
볼, 스트라이크, 낫싱을 결과 화면에 출력한다.

Lennon57 committed:
정답시 화면 출력 및 게임 재시작 버튼 활성화 및 기능 부여

Lennon57 committed:
테스트 통과 및 모든 기능 요구사항 구현
next_commit: 다른 요구사항 일치 확인

Lennon57 committed:
함수 명 변경 및 동작 변경
- 좀 더 직관적인 함수명으로 변경
- 함수 당 하나의 역할을 하도록 변경

pmhxhsj commented:
> 깃이 미숙해 테스트 commit이 많으며 11/28일부터 실제 커밋입니다. index.js 파일 하나에만 작성해야하는 줄 알았네요. 아직 객체지향이 미숙합니다.. 커밋도 많은 실수가 있는 것 같아요.. 거를 거 걸러주시고 조언 해주시면 감사드리겠습니다 ㅠㅠ 피드백듣고 다음 미션부터 적용하고 코드를 더더욱 발전시키겠습니다.

Tanney-102 commented on package.json:
```diff
@@ -2,7 +2,9 @@
   "description": "우아한테크코스 프리코스 숫자 야구 게임 미션",
-  "dependencies": {},
+  "dependencies": {
+    "@woowacourse/mission-utils": "^1.0.1"
```
모듈로 사용하는 거 좋네요👍

Tanney-102 replied:
다만 index.html에 mission-utils를 가져오는 script태그는 지우셔도 될 것 같아요!

Lennon57 replied:
경험이라 생각하고 두 가지 방법으로 시도해봤었는데 모듈로 사용했으면 script 태그를 빼는 것이 맞네요! 감사합니다 :)

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+
+const computerNumber = computerRandomNumber();
+
+function computerRandomNumber () {
```
함수이름에 함수의 동작도 명시해주면 어떨까요?
`getComputerRandomNumber`이런식으로요!

Lennon57 replied:
맞습니다 ㅠ 코딩 공부를 시작한 지 얼마 되지 않아 많이 미숙하네요! 함수명에 동작도 명시하는 거 꼭 기억하겠습니다!

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+const computerNumber = computerRandomNumber();
+
+function computerRandomNumber () {
+    let uniqueValueStack = [];
```
const를 사용해도 될 것 같습니다.
원시 타입과 객체 타입이 각각 식별자에 어떤 값을 할당하는지를 공부해보시면 좋을 것 같아요

Lennon57 replied:
넵! 말씀하신 부분 학습하겠습니다!

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    let uniqueValueStack = [];
+
+    while (uniqueValueStack.length < 3) {
+        let value = MissionUtils.Random.pickNumberInRange(1, 9);
```
여기도 const를 사용해도 될 것 같네요.
value라는 변수가 선언된 스코프 내에서 재할당되는 것처럼 보이지 않네요~

Tanney-102 replied:
더불어 3, 1, 9와 같은 상수들은 변수를 사용해 명시적으로 나타내면 좋을 것 같아요.
코드를 읽는 사람 입장에서 단순히 숫자만 봐서는 왜 3, 1, 9가 저 자리에 들어가는지 파악하기 어렵습니다.
예를들어 함수의 도메인이 `computerNumber`인 것 같으니
```js
const COMPUTER_NUMBER = {
    MIN: 1,
    MAX: 9,
    LENGTH: 3
}
```
이런식으로 선언해두고 사용하는 것이 좋겠네요.

Lennon57 replied:
앗 그러겠네요. 일일이 함수가 어떤 동작을 하는 지 검색해볼 수는 없으니까요. 다음부터 객체안에 선언하고 사용하는 게 코드 리뷰어한테 명확하게 다가올 수 있을 것 같네요!

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+        let value = MissionUtils.Random.pickNumberInRange(1, 9);
+        if (!uniqueValueStack.includes(value)) {
+            uniqueValueStack.push(value);
+        }
```
사소하긴한데 `uniqueValueStack`이라는 이름은 좀 안 어울리는 것 같아요!
전혀 스택처럼 행동하고 있지 않네요 ㅎㅎ 
단순 array로 사용할 목적이라면 `uniqueValues`와 같이 복수형으로 표시해주는 것만으로 충분합니다.

Tanney-102 replied:
추가로! 함수 안에서 중복되지 않은 숫자 3개가 뽑힐 때까지 while문을 돌리는 것 같은데 혹시 이 알고리즘의 시간복잡도를 예측할 수 있으신가요?? 
물론 확률론 관점에서 접근하면 충분히 수렴하는 값을 얻을 수는 있겠지만 일반적으로는 연산횟수를 특정하기 힘든 로직처럼 느껴지네요 ㅎㅎ 반복문의 끝을 예측할 수 없는 임의의 숫자들에 의존하고 있으니까요.
반복문을 사용하실 때는 가능한 그 시작과 끝을 온전히 통제하도록 사용하시는 걸 권장드립니다.

(아 물론, 해당 미션에서는 반복하는 size 자체가 작으니 크게 문제될 건 없어보이지만요!)

Lennon57 replied:
넵! 제가 생각해도 스택이란 이름은  어울리지 않네요.. 스택의 역할도 하지 않는데 왜 저렇게 지었는 지 모르겠네요. 연산횟수를 예측할 순 없습니다. 저도 리뷰어님의 생각처럼 사이즈가 3이니 쓰자! 라는 생각으로 했던 것 같아요 ㅎㅎ.. 다음 미션부터는 시간 복잡도를 계산할 수 있도록 반복문을 사용해보록 하겠습니다.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    return uniqueValueStack.join('');
+}
+
+function gameResultPrint (strikeNum, ballNum) {
```
여기도 함수이름에 동작이 포함되면 좋을 것 같네요! ㅎㅎ

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    };
+}
+
+import BaseBallGame from './index.js';
```
index.js안에서 index.js 모듈을 import 하나요?

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    return false;
+}
```
```js
return (
    input.length !== 3 || 
    isNaN(input) ||
    input.includes(0) || 
    input.includes(' ') ||
    [...new Set(input.split(''))].length !== input.length
);
```
로 충분하지 않을까요? 
각각 if문으로 그룹화하신 기준이 궁금합니다!

Tanney-102 replied:
```js
const validations = [
    input.length !== 3,
    isNaN(input),
    input.includes(0),
    input.includes(' '),
    [...new Set(input.split(''))].length !== input.length,
];

return validations.every(validation => validation);
```

이런 방법도 있겠네요

Lennon57 replied:
지금보니 그렇게 하는 게 가독성이 훨씬 좋네요! 여러 조건을 한 줄에 기입 했는데 컨벤션 규칙에 한 줄에 80글자가 넘으면 안된다는 거 보고 else if로 묶었던 것 같아요..  아래 every를 쓰는 방법은 생각치도 못한 방법이라 한 수 배우고 갑니다!

Tanney-102 replied:
eslint랑 prettier 고려해보시면 좋겠네요~ 컨벤션 규칙을 정해놓고 사용하시면 좋습니다.

Lennon57 replied:
앗 감사합니다! 한 번 알아보겠습니다

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    if (input === '3스트라이크') {
+        resultElement.innerHTML = "정답을 맞추셨습니다!<br>게임을 시작하시겠습니까?<button id='game-restart-button'>게임 재시작</button>";
+        document.querySelector('#game-restart-button').addEventListener('click', (e) => {
+            window.location.reload();
```
페이지 자체를 reload 하지 않고 초기화 하는 방법은 없을까요?
물론 해당 미션에서 이렇게 구현하게 문제는 없어 보이지만 그리 좋은 습관은 아닌 것 같습니다ㅎㅎ

Lennon57 replied:
헉 페이지 자체 reload 말고는 생각이 안 났는데 다시 생각해보니 input값 비우고, 함수를 실행해 시작하는 방법이 훨씬 좋은 것 같네요. 다음부터 동작을 생각하고 코드를 짜는 버릇을 길러야 겠어요

Tanney-102 replied:
input도 그렇고 computerNumber도 초기화 해주어야하겠네요~

Lennon57 replied:
넵 맞습니다! 다시 고쳐서 진행해보겠습니다!

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+}
+
+function resultPrintWindow () {
+    document.querySelector('#submit').addEventListener('click', (e) => {
```
버튼의 default 타입은 submit입니다.
즉, 기본적으로 버튼을 click하면 submit이벤트가 발생합니다.
이 버튼에 click이벤트를 달아주는 것과 form에 submit 이벤트를 달아서 처리를 해주는 것 사이에 차이점을 공부해보시면 좋을 것 같아요!
windowNotRefresh() 같은 유틸함수가 따로 필요한지 여부와 함께 고민해보시면 좋을 것 같습니다.

Lennon57 replied:
말씀하신 부분 한 번 공부해보도록 하겠습니다! 좋은 조언 감사드립니다 :)
말씀하신 부분 찾아보니 유틸함수 대신 form자체에 refresh되지 않게하는 기능이 있는 것 같습니다 (onsubmit="return false")

Tanney-102 replied:
음 onsubmit = "return false" 도 꼭 필요할걸까요??
애초에 submit이벤트를 발생시키지 않을 목적이라면 차라리 버튼에 type="button" 어트리뷰트를 주는 게 낫습니다.
제가 중점적으로 말씀드리고 싶었던 건 submit 이벤트를 '잘 활용하는' 방법입니다!

form태그 안에서 버튼의 역할은 무엇일까요?
물론 구현해야하는 대상에 따라 단순 클릭이 목적일 수 있지만 form의 버튼은 submit을 목적으로 합니다. 실제로 type="submit"이 default기도 하구요.
따라서 click 이벤트보다는 submit이벤트에 핸들러를 달아주는 더 적절하지 않나하는 생각이 들어서 드린 피드백이었습니다.

목적성을 떠나서 기능적으로도 그게 더 자연스럽긴 합니다.
예를들어 지금 구현하신 코드에 'enter' 키로 숫자를 제출하는 기능을 추가한다면 키보드 관련 이벤트를 또 잡아줘야할까요?
input에서 'enter'를 누르든 form안에 type="submit" 버튼을 누르든 submit이벤트가 발생하니 본래 목적에 맞는 submit 이벤트에 핸들러를 달아주는게 더 자연스럽죠~

자세한건 공부해보시면 좋을 것 같구요. 간단한 예시 두 가지정도 첨부합니다.

1. submit 이벤트 활용
```js
// index.html
<form class="js-form">
    <input type="text" />
    <button>click</button>
</form>

// index.js
document.querySelector(".js-form").addEventListener("submit", (event) => {
    event.preventDefault();

    // logic....
});
```

2. submit없이 click이벤트만을 사용
```js
// index.html
<form class="js-form">
    <input type="text" />
    <button type="button">click</button>
</form>

// index.js
document.querySelector(".js-form").addEventListener("click", (event) => {
    // logic....
});
```

Lennon57 replied:
말씀하신 대로 js를 통해 핸들링하는 방법이 더 좋을 것 같아요. 유틸함수 대신 그냥 내용값만 지정해주는 게 더 가독성이 좋을 것 같습니다! 제가 리뷰어님이 처음 말씀하신 내용을 잘못 이해했네요.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+// (ㅇ) 입력한 세 숫자에 따라 볼, 스트라이크, 낫싱 출력(볼이 스트라이크보다 앞에 출력)
+// (ㅇ) 정답 화면 및 재시작 버튼 출력
+
+const computerNumber = computerRandomNumber();
```
BaseballGame클래스의 핵심 멤버처럼 보이는데 전역변수로 관리하는게 좋을까요?

Lennon57 replied:
음... 저도 이 부분이 가장 문제였는데요. 전역변수로 선언하지 않으면 play함수가 실행될 때마다 컴퓨터 랜덤값이 값이 계속 변해서 게임 진행 자체가 되지 않았습니다. 
제가 객체지향을 잘 이해하지 못해서 그런 것 같아요.. 혹시 이 부분에 대해서 한 가지 조언 해주실 수 있을까요?? 
미션을 PR하고 몇 번 더 시도 해봤는데 아직도 방법을 못 찾았습니다.

Tanney-102 replied:
숫자 야구 게임의 진행자를 하나의 객체로 생각해봅시다!
그리고 그 진행자가 '무엇을 알고 있어야 하는지', '무엇을 할 수 있어야 하는지'를 고민해봅시다ㅎㅎ

예를들면 이렇게 표현할 수 있겠죠.
숫자 야구 진행자 {
    알고 있는 것: 컴퓨터가 고른 랜덤 숫자
    할 수 있는 것: 컴퓨터 숫자 고르기, 게임 진행, 게임 초기화....
}

숫자 야구 진행자 => BaseballGame 클래스
컴퓨터 숫자 => computerNumber

...
우선 이 정도까지 힌트를 드리겠습니다. 고민해보시고 더 이야기 나눠봤으면 좋겠네요!

Lennon57 replied:
감사합니다! 한 번 깊게 고민해보고 코드 작성해보도록 하겠습니다

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,111 @@
+    document.querySelector('#submit').addEventListener('click', (e) => {
+
+        let inputValue = document.querySelector('#user-input').value;
+        let resultString = new BaseBallGame().play(computerNumber, inputValue.split(''));
```
이렇게 사용한다면 BaseballGame이라는 클래스가 굳이 존재할 이유가 없어보여요 ㅎㅎ
그냥 play 함수만 있는게 더 간결하고 의미있는 코드가 아닐까요?

단순히 미션에서 BaseballGaem 클래스를 제시했기때문에 사용하는 건지, 혹은 특정 상태와 그 동작을 객체로 관리하기 위해 사용하는 건지 분명히하시면 좋을 것 같습니다.

Lennon57 replied:
맞습니다. 이 부분도 크게 고민이 된 부분입니다. 말씀하신대로 play함수만 있는 게 더 간결하고 의미있습니다. 미션에서 export, import를 요구하는 것 같은데, 저는 그저 index.js에서만 진행해서 필요성이 없을 거라 생각했어요.. 그런데도 뭔가 사용해야 될 것 같아서 사용했네요. export, import의 기능에 대해 더 파악하고 말씀하신 부분을 분명히 해야할 필요성이 있을 것 같습니다.

지금까지 제 미련한 코드를 리뷰해주셔셔 감사합니다. 정말 많은 도움이 된 것 같아요! 
다음 미션은 말씀하신 부분을 고쳐 훨씬 더 좋고 명확한 코드로 작성하도록 하겠습니다. 
코드리뷰 열어주셔셔 정말 감사합니다. 좋은 하루 보내세요!

Tanney-102 replied:
돌아가면 다 좋은 코드 아니겠습니까~ 
미련한지 아닌지는 나중에 성장하고 나서 돌아볼 때 판단하는 걸로!

Tanney-102 reviewed (changes requested):
안녕하세요. 우아한테크코스 3기 프론트엔드 브콜입니다.ㅎㅎ
빠르게 미션 구현해주셨네요!
몇가지 피드백 드립니다.

1. ES6 모듈에 대해 더 공부하실 것을 추천드립니다.
2. 기본적으로 식별자의 이름을 최대한 명시적으로 나타내려고 하신 점이 정말 좋았습니다! 다만 함수 이름만큼은 그 함수의 '동작'(동사)도 함께 명시해주면 더욱 좋을 것 같아요!
3. const와 let, 원시타입과 객체 타입에 대해 더 공부해보시면 좋을 것 같습니다.
4. 객체지향 프로그래밍에 대해 더 공부하실 필요가 있습니다.

더 자세한 내용은 코멘트로 남겼습니다.

Tanney-102 commented:
리뷰도 빠르게 확인해주셨네요! ㅎㅎ
혹시 코멘트 드린 것을 바탕으로 수정해보고 싶으시다면 수정하시도 같은 브랜치로 push 하시면 됩니다.

Lennon57 commented:
> 리뷰도 빠르게 확인해주셨네요! ㅎㅎ 혹시 코멘트 드린 것을 바탕으로 수정해보고 싶으시다면 수정하시도 같은 브랜치로 push 하시면 됩니다.

감사합니다 ㅠㅠ 피드백을 수용해 코드를 수정하고 push 해보도록 하겠습니다! 바쁘신데 고생이 많으십니다 :)

Lennon57 committed:
eslint, prettier을 적용했습니다.
- airbnb-base를 적용했습니다.

함수명 변경
- 함수이름에 동작이 포함되도록 하였습니다.

전역 변수(computerNumber)
- 함수(baseBallGame)로 게임 실행을 하도록 동작하여 해결하였습니다.

export 제외
- 하나의 js파일에서 진행해 import-export를 제외해 가독성을 높혔습니다.
(다음 미션부터는 사용할 예정입니다.))

게임 재시작
- reload하는 것이 아닌 버튼을 누르면baseBallGame함수가 실행되도록 변경하였습니다.

missoion-utils 모듈(x)
- 계속 오류가 떠 cdn으로 유지했습니다.

새로고침 유틸 함수 사용하지 않기
- 유틸 함수 대신 js에서 form에 핸들링하여 실행되도록 변경하였습니다.

컴퓨터 랜덤함수 3개 시간복잡도(x)
- 랜덤으로 뽑아야해 어쩔 수 없이 while문을 사용하고 배열대신 set을 이용해 명확하게 하였습니다.

const, let 수정
- 스코프 내에서 재할당 되지 않는 변수들을 const로 선언하였습니다.

현재 너무 바빠 많은 시간을 할애하지 못해 여기까지가 이번 주 미션의 마무리일 것 같습니다.
다음 주부터는 export, import를 사용하여 js파일을 나눠 각 파일마다 역할을 줘 명확하게 보이는 것이 목표입니다.
코드 리뷰해주셔셔 감사합니다!
많은 도움이 됐습니다 :)
